### PR TITLE
Separate settings files for production, development and testing.

### LINKED ISSUE #5: Distinguish development and production setup
- [ ] Understand the best way to do this (have settings for each setup & being able to easily use them)
- [ ] Database
  - [X] development sqlite
  - [ ] production postgres (there is a how to on running postgres using docker from the original backend prototype, here just need to add settings to connect to the db)
- [ ] Authentication (right now all methods are in the settings, so it would development
  - [ ] production: 'rest_framework_simplejwt.authentication.JWTAuthentication'
  - [ ] development
    - 'rest_framework_simplejwt.authentication.JWTAuthentication'
    - 'rest_framework.authentication.BasicAuthentication'
    - 'rest_framework.authentication.SessionAuthentication'

### PR DESCRIPTION
**Pull request by sgarciahelguera**
Mainly remove auth alternatives for production and testing, and add postgresql configuration for production.

Close #5

### CODE REVIEW

File: cubeseed/settings_test.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+DATABASES = {
+    "default": {
+        "ENGINE": "django.db.backends.sqlite3",
+        "NAME": BASE_DIR / "db.sqlite3",
+    }
+}
+
+REST_FRAMEWORK["TEST_REQUEST_DEFAULT_FORMAT"] = "json"
+REST_FRAMEWORK["TEST_REQUEST_RENDERER_CLASSES"] = [
+    "rest_framework.renderers.JSONRenderer",
```
Comment (Natiman58) [inline cubeseed/settings_test.py:16]: how about handling file uploads of multipart files during testing ; may be can add Multipart renderer
"rest_framework.renderers.MultiPartRenderer" into our framework

Comment (sgarciahelguera) [reply cubeseed/settings_test.py:16]: Just merged from main and it looks like you added that alongside some test, right? 
It is now on test and development settings.

File: docker/docker-compose-dev.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      dockerfile: docker/Dockerfile.local
+      context: ..
+      args:
+        - DJANGO_SETTINGS_MODULE=cubeseed.settings
+    environment:
+      - DB_NAME=${DB_NAME}
+      - DB_USER=${DB_USER}
+      - DB_PASS=${DB_PASS}
+      - DB_HOST=db
+    command: manage.py runserver 0.0.0.0:8000
```
Comment (mihrab34) [inline docker/docker-compose-dev.yml:17]: can you add a `.env.example` file so contributors can know what to include in their env files when running the docker container

**Comment by Reid-T-W** [comment]
Looks great, just added one minor comment

File: docker/docker-compose-dev.yml (hunk 2 above)
Comment (mihrab34) [reply docker/docker-compose-dev.yml:17]: Testing the auth token endpoint on localhost also gave cors error You might want to look into that

cors error

File: cubeseed/settings.py
[the remark this answers is not in the document]
Comment (sgarciahelguera) [reply cubeseed/settings.py:1]: added 127.0.0.1 to allowed hosts

File: docker/docker-compose-dev.yml (hunk 2 above)
Comment (sgarciahelguera) [reply docker/docker-compose-dev.yml]: Added `docker/.env` with the basics (enough for development).

**Comment by sgarciahelguera** [comment]
Ended up moving `docker/docker-compose-dev.yml` to `docker-compose.yml` as having a symbolic link was not trivial (context has to point to the root of the project, relatively, so if you run from docker folder is one thing but if you run from a symlink in some other directory it has to be something else).

**Comment by sgarciahelguera** [comment]
Everything should be working now for development, I'll leave some work on production and probably removing the testing settings for another pull request.
Please review/approve with that in mind.

**Comment by Reid-T-W** [comment]
When accessing the swagger endpoint it loads up but there is still some issue with the FarmPlanner, kindly take a look at it below.

https://docs.google.com/document/d/1T5uePXzeNDF4L3-_YAwyO2KQgcUKOrybr5ttwtaSuwo/edit?usp=sharing

The exact steps I followed are

1. Clone the issue#5 branch 
2. Added .env file
3. Run `docker-compose up`
4. Access the "http://localhost:8000/swagger/" enpoint

**Comment by Reid-T-W** [comment]
When accessing the (GET) /userauth/groups/ endpoint i get the following error, kindly take a look via this link

https://docs.google.com/document/d/1WjOPYxjvWwAC4TicvSGyfkg62Hy9F1oQos3Q4PGftn4/edit?usp=sharing

The exact steps I followed are

Clone the issue#5 branch
Added .env file
Run docker-compose up
Access the (GET) /userauth/groups/ endpoint via swagger

**Comment by Reid-T-W** [comment]
I also noticed that there is a workaround put in place to handle the FarmPlanner and Swagger issue, the swagger loads up but I'm still facing errors as shown in this link

https://docs.google.com/document/d/1T5uePXzeNDF4L3-_YAwyO2KQgcUKOrybr5ttwtaSuwo/edit?usp=sharing

The exact steps I followed are

Clone the issue#5 branch
Added .env file
Run docker-compose up
Access the "http://localhost:8000/swagger/" endpoint

**Comment by Reid-T-W** [comment]
> When accessing the (GET) /userauth/groups/ endpoint i get the following error, kindly take a look via this link
> 
> https://docs.google.com/document/d/1WjOPYxjvWwAC4TicvSGyfkg62Hy9F1oQos3Q4PGftn4/edit?usp=sharing
> 
> The exact steps I followed are
> 
> Clone the issue#5 branch Added .env file Run docker-compose up Access the (GET) /userauth/groups/ endpoint via swagger

As per our conversation I tried to manually export the env variables, but the issue persisted. But it seems like manually running the python3 manage.py migrate command in the "cubeseed-backend-dev" container solves the issue.

**Comment by sgarciahelguera** [comment]
> > When accessing the (GET) /userauth/groups/ endpoint i get the following error, kindly take a look via this link
> > https://docs.google.com/document/d/1WjOPYxjvWwAC4TicvSGyfkg62Hy9F1oQos3Q4PGftn4/edit?usp=sharing
> > The exact steps I followed are
> > Clone the issue#5 branch Added .env file Run docker-compose up Access the (GET) /userauth/groups/ endpoint via swagger
> 
> As per our conversation, I tried to export the env variables manually, but the issue persisted. However, the issue disappeared when the "python3 manage.py migrate" was run in the "cubeseed-backend-dev" container manually.

migration is run as a precondition, it should be done every time you update 
Maybe you had some changes come up and didn't update the container? 
Next time you see something like it try to do `docker-compose -f docker/docker-compose-dev.yml up --build`  (or with the corresponding yml file), and if it still fails try adding a verbose flag, it may add some useful info (or just a lot of noise) : `docker-compose --verbose -f docker/docker-compose.yml up --build`

**Comment by sgarciahelguera** [comment]
I just removed gunicorn, went back to django's runserver. I need to solve static files for gunicorn to work properly, I will do that for production together with testing AWS cloud settings.

I think I mentioned it before, but the entrypoint for the docker container is now a script that loads environment from a `.env` file if present, so that should solve some previous issues and simplify things.

**Comment by Reid-T-W** [comment]
> > > When accessing the (GET) /userauth/groups/ endpoint i get the following error, kindly take a look via this link
> > > https://docs.google.com/document/d/1WjOPYxjvWwAC4TicvSGyfkg62Hy9F1oQos3Q4PGftn4/edit?usp=sharing
> > > The exact steps I followed are
> > > Clone the issue#5 branch Added .env file Run docker-compose up Access the (GET) /userauth/groups/ endpoint via swagger
> > 
> > 
> > As per our conversation, I tried to export the env variables manually, but the issue persisted. However, the issue disappeared when the "python3 manage.py migrate" was run in the "cubeseed-backend-dev" container manually.
> 
> migration is run as a precondition, it should be done every time you update Maybe you had some changes come up and didn't update the container? Next time you see something like it try to do `docker-compose -f docker/docker-compose.yml up --build` (or with the corresponding yml file), and if it still fails try adding a verbose flag, it may add some useful info (or just a lot of noise) : `docker-compose --verbose -f docker/docker-compose.yml up --build`

I didn't make any changes, but to be on the safe side I recloned the repo (branch issue#5) and run the command you suggested `docker-compose --verbose -f docker/docker-compose.yml up --build` but the issue persists. I understand that migration is being run as a precondition in a separate container, but it seems like the backend container is not accessing those migrations, can the issue be with the volume these containers are sharing?

**Comment by sgarciahelguera** [comment]
OK, I was trying `docker/docker-compose.yml` and suspect you were trying the one in the root.
The issue is exactly what you say, but not because of a shared volume, sqlite lives on the local filesystem of the container.
So the migrations are applied in a container, not in the other.

**Comment by Reid-T-W** [comment]
> OK, I was trying `docker/docker-compose.yml` and suspect you were trying the one in the root. The issue is exactly what you say, but not because of a shared volume, sqlite lives on the local filesystem of the container. So the migrations are applied in a container, not in the other.

Okay, that's a valid point, I think we can raise this as a separate issue and merge the changes, should we do that?

**Comment by sgarciahelguera** [comment]
I moved the migrate command into the runner script (`docker/runner.sh`) and removed migration from docker compose files.
I did a quick run for both `docker/docker-compose.yml` and `docker-compose.yml` and it looks like they are running.
docker compose from github built images is not tested yet, I'll leave it for later.

**Comment by Reid-T-W** [comment]
Great, it works on my end as well.

**Review by Reid-T-W — APPROVED** [review]

**Pull request merged**